File: java/AlgorithmAndDataStructure/src/main/java/algorithms/Sorting.java
package algorithms;

import java.util.*;

/**
 * Created by lzx on 15-2-19.
 */
public class Sorting {
	private static <T> void swap(List<T> list, int a, int b) {
		T tmp = list.get(a);
		list.set(a, list.get(b));
		list.set(b, tmp);
	}

	public static <T extends Comparable> boolean isSorted(List<T> a) {
		int size = a.size();
		for (int i = 1; i < size; i++) {
			if (a.get(i - 1).compareTo(a.get(i)) > 0)
				return false;
		}
		return true;
	}

	public static <T extends Comparable> ListIterator<T> isSortedUntil(List<T>
	a) {
		ListIterator<T> i = a.listIterator();
		if (!i.hasNext())
			return i;
		T prev = i.next();

		while (i.hasNext()) {
			T cur = i.next();
			if (prev.compareTo(cur) > 0) {
				i.previous();
				return i;
			}
			prev = cur;
		}
		return i;
	}

	// Quicksort
	public static <T extends Comparable> void sort(List<T> a) {
		quicksort(a, 0, a.size() - 1);
	}

	private static <T extends Comparable> void quicksort(List<T> a, int
			start, int end) {
		if (start < end) {
			int pivot = partition(a, start, end);
			quicksort(a, start, pivot - 1);
			quicksort(a, pivot + 1, end);
		}
	}

	private static <T extends Comparable> int partition(List<T> a, int start,
	                                                    int end) {
		T x = a.get(end);
		int i = start - 1;
		for (int j = start; j < end; j++) {
			if (a.get(j).compareTo(x) <= 0) {
				i += 1;
				swap(a, i, j);
			}
		}
		swap(a, i + 1, end);
		return i + 1;
	}

	public static <T extends Comparable> void partialSort(List<T> a, int
			border) {
		ArrayList<T> tmp = new ArrayList<T>();
		Heap.makeHeap(a);
		int bound = a.size();
		for (int i = 0; i < border && i < bound; i++) {
			tmp.add(Heap.popHeap(a));
		}
		int size = tmp.size();
		for (int i = size - 1; i >= 0; i--) {
			a.add(0, tmp.get(i));
		}
	}

	public static <T extends Comparable> void partialSortCopy(List<T> a, int
			border, List<T> b) {
		ArrayList<T> tmp = new ArrayList<T>(a);
		Heap.makeHeap(tmp);
		int bound = a.size();
		for (int i = 0; i < border && i < bound; i++) {
			b.add(Heap.popHeap(tmp));
		}
	}

	// Heap.make_heap will construct a min-heap, so we use it to
	// find the nlargest element, which n is one-base.
	// assume 0 < n < a.size()
	public static <T extends Comparable> T nthElement(List<T> a, int n) {
		ArrayList<T> tmp = new ArrayList<T>(a);
		Heap.makeHeap(tmp);
		int size = tmp.size();
		for (int i = n; i < size; i++) {
			if (a.get(i).compareTo(tmp.get(0)) > 0) {
				Heap.popHeap(tmp);
				Heap.pushHeap(tmp, a.get(i));
			}
		}
		return Heap.popHeap(tmp);
	}

	public static <T> void stableSort(List<T> a, Comparator<T> cp) {
		int size = a.size();
		if (size <= 1)
			return;
		ArrayList<Integer> pos = new ArrayList<Integer>(size);
		for (int i = 0; i < size; i++) {
			pos.add(i);
		}
		stableQuicksort(a, 0, size - 1, cp, pos);

		int from = 0;
		for (int to = 1; to < size; to++) {
			if (cp.compare(a.get(from), a.get(to)) != 0) {
				if (to - from > 1) {
					Collections.sort(a.subList(from, to), cp);
				}
				from = to;
			}
		}
	}

	private static <T> void stableQuicksort(List<T> a, int
			start, int end, Comparator<T> cp, ArrayList<Integer> pos) {
		if (start < end) {
			int pivot = stablePartition(a, start, end, cp, pos);
			stableQuicksort(a, start, pivot - 1, cp, pos);
			stableQuicksort(a, pivot + 1, end, cp, pos);
		}
	}

	private static <T> int stablePartition(List<T> a, int
			start, int end, Comparator<T> cp, ArrayList<Integer> pos) {
		T x = a.get(end);
		int i = start - 1;
		for (int j = start; j < end; j++) {
			if (cp.compare(a.get(j), x) <= 0) {
				i += 1;
				swap(a, i, j);
				swap(pos, i, j);
			}
		}
		swap(a, i + 1, end);
		swap(pos, i + 1, end);
		return i + 1;
	}
}

File: scala/src/main/scala/algorithm/Sorting.scala
package algorithm

import collection.mutable.Buffer
import collection.mutable.Seq
import collection.mutable.IndexedSeq

object Sorting {
  // a compare function used by quicksort
  private def compare[T](a: T, b: T)(implicit ord: math.Ordering[T])  = {
    if (ord.equiv(a, b)) {
      0
    } 
    else {
      if (ord.gt(a, b)) 1 else -1
    }
  }

  //  a helper function which swap the value in two position of a mutable Seq
  private def swap[T](list: Seq[T], a: Int, b: Int) {
    val tmp = list(a)
    list(a) = list(b)
    list(b) = tmp
  }

  def IsSorted[T](c: Seq[T])(implicit ord: math.Ordering[T]) : Boolean = {
    val size = c.size
    for (i <- 1 until size) {
      if (ord.gt(c(i - 1), c(i))) {
        return false
      }
    }
    return true
  }

  def IsSortedUntil[T](c: Seq[T])(implicit ord: math.Ordering[T]) : Int = {
    val size = c.size
    for (i <- 1 until size) {
      if (ord.gt(c(i - 1), c(i))) {
        return i
      }
    }
    return size
  }

  def Sort[T](c: Seq[T])(implicit ord: math.Ordering[T]) {
    quicksort(c, 0, c.size - 1, compare[T])
  }

  private def quicksort[T](c: Seq[T], start: Int, 
    end: Int, cp: (T, T) => Int)(implicit ord: math.Ordering[T]) {

    if (start < end) {
      val pivot = partition(c, start, end, cp)
      quicksort(c, start, pivot - 1, cp)
      quicksort(c, pivot + 1, end, cp)
    }
  }

  private def partition[T](c: Seq[T], start: Int, 
    end: Int, cp: (T, T) => Int)(implicit ord: math.Ordering[T]) = {
    
    val sential = c(end)
    var i = start - 1
    for (j <- start until end) {
      if (cp(sential, c(j)) > 0) {
        i += 1
        swap(c, i, j)
      }
    }
    swap(c, i + 1, end)
    i + 1
  }

  def StableSort[T](c: Seq[T], 
    cp: (T, T) => Int)(implicit ord: math.Ordering[T]): Unit = {
    
    val size = c.size
    if (size <= 1)
      return 
    var pos = Seq.iterate(0, size)(x => x + 1)
    stableQuicksort(c, 0, size - 1, cp, pos)

    var from = 0
    for (to <- 1 until size) {
      if (cp(c(from), c(to)) != 0) {
        if (to - from > 1) { // optimization
          stableQuicksort(pos, from, to - 1, (x: Int, y: Int) => {
            if (x > y) 1 else 0
          }, c)
        }
        from = to
      }
    }
  }

  private def stableQuicksort[T, E](c: Seq[T], start: Int, end: Int, 
    cp: (T, T) => Int, pos: Seq[E])(implicit ord: math.Ordering[T]) {
      
    if (start < end) {
      val pivot = stablePartition(c, start, end, cp, pos)
      stableQuicksort(c, start, pivot - 1, cp, pos)
      stableQuicksort(c, pivot + 1, end, cp, pos)
    }
  }

  private def stablePartition[T, E](c: Seq[T], start: Int, end: Int, 
    cp: (T, T) => Int, pos: Seq[E])(implicit ord: math.Ordering[T]) = {
    
    val sential = c(end)
    var i = start - 1
    for (j <- start until end) {
      if (cp(c(j), sential) <= 0) {
        i += 1
        swap(c, i, j)
        swap(pos, i, j)
      }
    }
    swap(c, i + 1, end)
    swap(pos, i + 1, end)
    i + 1
  }

  def PartialSort[T](c: Seq[T],
    border: Int)(implicit ord: math.Ordering[T]) = {
    
    var tmp = c.toBuffer
    Heap.MakeHeap(tmp)
    var bound = border min c.size
    for (i <- 0 until bound) yield Heap.PopHeap(tmp)
  }

  // No need to implement PartialSortCopy
  
  // Elements in [0, n) are all less than element in n
  def NthElement[T](c: Seq[T], 
    n: Int)(implicit ord: math.Ordering[T]) : Unit = {
    
    var left = 0
    var right = c.size - 1
    while (true) {
      if (left >= n || right <= n) {
        return
      }
      val nth = quickselect(c, left, right, n)
      if (nth == n) {
        return 
      }
      if (nth < n) {
        left = nth
      }
      else {
        right = nth
      }
    }
  }

  private def quickselect[T](c: Seq[T], left: Int, right: Int, 
    border: Int)(implicit ord: math.Ordering[T]) = {
    
    swap(c, border, right)
    var idx = left
    for (i <- left until right) {
      if (ord.lt(c(i), c(right))) {
        swap(c, idx, i)
        idx += 1
      }
    }
    swap(c, idx, right)
    idx
  }
}
